core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.docs;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.internal.common.JacksonUtil;
+
+/**
+ * Generates a JSON Schema from the given service specification.
+ */
+@UnstableApi
+public final class JsonSchemaGenerator {
+
+    private JsonSchemaGenerator() {
+        throw new UnsupportedOperationException();
+    }
+
+    private static final ObjectMapper mapper
+            = JacksonUtil.newDefaultObjectMapper();
+
+    private static final class FieldSchemaWithAdditionalProperties {
+        private final ObjectNode node;
+        private final Set<String> requiredFieldNames;
+
+        private FieldSchemaWithAdditionalProperties(ObjectNode node, Set<String> requiredFieldNames) {
+            this.node = node;
+            this.requiredFieldNames = requiredFieldNames;
+        }
+    }
+
+    private static String getSchemaType(TypeSignature type) {
+        if (type.isNamed()) {
+            // Hacky way because protobuf library is not in classpath.
+            // We can consider moving JSONSchemaGenerator for protos to grpc or protobuf lib.
+            if (type.namedTypeDescriptor() != null && type.namedTypeDescriptor().getClass().getName().contains(
+                    "Enum")) {
+                return "string";
+            } else {
+                return "object";
+            }
+        }
+
+        if (type.isContainer()) {
+            switch (type.name().toLowerCase()) {
+                case "repeated":
+                case "list":
+                case "array":
+                    return "array";
+                default:
```

Comment:
I believe we use `list`, `set` and `map` for the container type.

Reply:
For `map` type, I think easiest thing to do is to just convert it to an `object`.

Reply:
Resolving because types signatures contain this info now.